infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/PVCSubPathHelper.java

```diff
@@ -247,6 +261,56 @@ private void execute(
     }
   }
 
+  /**
+   * Watches failure events for the specified Pod, and deletes the Pod if a failure event is
+   * detected
+   *
+   * @param deployments the KubernetesDeployments object
+   * @param podName the name of the Pod to check events for
+   * @throws InfrastructureException
+   */
+  private void watchFailureEventAndTerminatePod(KubernetesDeployments deployments, String podName)
+      throws InfrastructureException {
+    deployments.watchEvents(
+        event -> {
+          if (podName.equals(event.getPodName()) && isPodFailureEvent(event)) {
+            try {
+              LOG.debug(
+                  "Deleting pod: '{}' due to failure event: '{}'", podName, event.getMessage());
+              deployments.delete(event.getPodName());
+            } catch (InfrastructureException ex) {
+              LOG.error("Unable to delete failing pod: '{}' cause: '{}'", podName, ex.getMessage());
+            }
+          }
+        });
+  }
+
+  private boolean isPodFailureEvent(PodEvent event) {
+    return event.getReason().equals(POD_EVENT_FAILED)
+        || event.getReason().equals(POD_EVENT_FAILED_SCHEDULING)
+        || event.getReason().equals(POD_EVENT_FAILED_MOUNT);
+  }
```

Comment:
This has some overlap with [`UnrecoverablePodEventListenerFactory`](https://github.com/eclipse-che/che-server/blob/main/infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/util/UnrecoverablePodEventListenerFactory.java) -- I wonder if there's some way to reuse the functionality (maybe not, since the unrecoverable events there are configured by property)

Reply:
As far as I recall, unrecorerable events kick in only during the workspace startup phase, async removal of dirs in PVC is a separate routine

Reply:
Also, one of my concerns about watching for these failure events is that: a terminating PVC would cause failure events for the rm-pods, but, observing failure events might not necessarily mean that its caused by a terminating PVC? Maybe there could be other reasons (that are not PVC-related) why these rm pods might have failure events?